File: LeetCode/141. Linked List Cycle/Solution.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) {
 * val = x;
 * next = null;
 * }
 * }
 */
public class Solution {
    //https://en.wikipedia.org/wiki/Cycle_detection#Tortoise_and_hare
    //Floyd's cycle-finding algorithm
    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null || head.next.next == null) {
            return false;
        }
        ListNode slow = head.next, fast = head.next.next;
        while (slow != fast) {
            if (fast.next == null || fast.next.next == null) {
                return false;
            }
            fast = fast.next.next;
            slow = slow.next;
        }
        return true;
    }
}

File: LeetCode/141. Linked List Cycle/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head or not head.next:
            return False
        slow, fast = head.next, head.next.next
        while fast:
            if slow is fast:
                return True
            slow = slow.next
            if not fast.next or not fast.next.next:
                return False
            fast = fast.next.next
        return False

    def hasCycle(self, head):
        try:
            slow, fast = head.next, head.next.next
            while slow is not fast:
                slow = slow.next
                fast = fast.next.next
            return True
        except Exception:
            return False
